source/framework/vx_tensor.c

```diff
@@ -28,21 +28,127 @@
 Tensor HELPER FUNCTIONS
 =============================================================================*/
 
+static vx_status isTensorSwappable(vx_reference input, vx_reference output);
+static vx_status moveOrSwapTensor(vx_reference input, vx_reference output);
+static vx_status VX_CALLBACK tensorKernelCallback(vx_enum kernel_enum, vx_bool validate_only, const vx_reference params[], vx_uint32 num_params);
+static vx_bool ownIsValidTensorFormat(vx_enum data_type);
 static void ownInitTensorObject(
     vx_tensor tensor, const vx_size* dimensions, vx_size number_of_dimensions, vx_enum data_type, vx_int8 fixed_point_position);
-static vx_bool ownIsValidTensorFormat(vx_enum data_type);
 static vx_status ownTensorCheckSizes(const volatile uint32_t *dimensions, const vx_size * view_start, const vx_size * view_end, vx_size number_of_dimensions);
 static vx_size ownComputePatchSize (const vx_size * view_start, const vx_size * view_end, vx_size number_of_dimensions);
 static void ownComputePositionsFromIndex(vx_size idx, const vx_size * start, const vx_size * end,
         const volatile uint32_t * tensor_stride, const vx_size * patch_stride,  vx_size number_of_dimensions,
         vx_size * tensor_pos, vx_size * patch_pos);
 static vx_uint32 ownComputePatchOffset(vx_size num_dims, const vx_size *dim_coordinate, const volatile uint32_t *strides);
 
-static vx_bool ownIsValidTensorFormat(vx_enum data_type)
+
+/*! \brief check to see if the tensors may be swapped.
+ * They must be copyable, and not sub-tensors
+*/
+static vx_status isTensorSwappable(vx_reference input, vx_reference output)
+{
+    vx_status status = (vx_status)VX_SUCCESS;
+    if ((vx_bool) vx_true_e ==  tivxIsReferenceMetaFormatEqual(input, output))
+    {
+        tivx_obj_desc_tensor_t * ip_obj_desc = (tivx_obj_desc_tensor_t *)input->obj_desc;
+        tivx_obj_desc_tensor_t * op_obj_desc = (tivx_obj_desc_tensor_t *)output->obj_desc;
+        if (ip_obj_desc->mem_size != op_obj_desc->mem_size)
+        {
+           status = (vx_status)VX_ERROR_NOT_SUPPORTED;
+       }
+    }
+    else
+    {
+        status = (vx_status)VX_ERROR_NOT_SUPPORTED;
+    }
+    return status;
+}
+
+/*! \brief Swap orMove the input tensor to the output
+ * Note this is the same as swap if the input is not virtual
+ */
+static vx_status moveOrSwapTensor(vx_reference input, vx_reference output)
 {
-    vx_bool res = (vx_bool)vx_false_e;
+    vx_status status =  ownReferenceLock(output);
+    if ((vx_status)VX_SUCCESS == status)
+    {
+        /* swap destructors */
+        tivx_reference_callback_f destructor = output->destructor_callback;
+        output->destructor_callback = input->destructor_callback;
```

Comment:
Why do we need to swap the destructor callbacks ... for tensor, they should be the same.

Reply:
@raphacan I reviewed this and while these are generic callbacks, it is fine to leave the code as/is.  I'd suggest though adding a comment to say that this has been added for completeness and to ensure that in case there is ever a need for unique destructors, this has been accounted for

Reply:
done